vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+				logical.UpdateOperation: i.pathOIDCCreateUpdateRole,
+				logical.CreateOperation: i.pathOIDCCreateUpdateRole,
+				logical.ReadOperation:   i.pathOIDCReadRole,
+				logical.DeleteOperation: i.pathOIDCDeleteRole,
+			},
+			ExistenceCheck: i.pathOIDCRoleExistenceCheck,
+		},
+		{
+			Pattern: "oidc/role/?$",
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ListOperation: i.pathOIDCListRole,
+			},
+			HelpSynopsis:    "list oidc/role/ help synopsis here",
+			HelpDescription: "list oidc/role/ help description here",
+		},
+		{
+			Pattern: "oidc/introspect/?$",
+			Fields: map[string]*framework.FieldSchema{
+				"token": {
+					Type:        framework.TypeString,
+					Description: "Token to verify",
+				},
+				"client_id": {
+					Type:        framework.TypeString,
+					Description: "Optional client_id to verify",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: i.pathOIDCIntrospect,
+			},
+			HelpSynopsis:    "list oidc/introspect/ help synopsis here",
+			HelpDescription: "list oidc/introspect/ help description here",
+		},
+	}
+}
+
+func (i *IdentityStore) pathOIDCReadConfig(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	c, err := i.getOIDCConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	if c == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"issuer": c.Issuer,
+		},
+	}, nil
+}
+
+func (i *IdentityStore) pathOIDCUpdateConfig(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	value, ok := d.GetOk("issuer")
+	if !ok {
+		return nil, nil
+	}
+
+	c := oidcConfig{
+		Issuer: value.(string),
+	}
+
+	entry, err := logical.StorageEntryJSON(oidcConfigStorageKey, c)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	var resp logical.Response
+
+	if c.Issuer != "" {
+		resp.AddWarning(`If "issuer" is set explicitly, all tokens must be ` +
+			`validated against that address, including those issued by secondary ` +
+			`clusters. Setting issuer to "" will restore the default behavior of ` +
+			`using the cluster's api_addr as the issuer.`)
+	}
+
+	i.oidcCache.purge()
+
+	return &resp, nil
+}
+
+func (i *IdentityStore) getOIDCConfig(ctx context.Context, s logical.Storage) (*oidcConfig, error) {
+	lock := i.oidcCache.lock
+
+	lock.RLock()
+	if i.oidcCache.config != nil {
+		defer lock.RUnlock()
+		return i.oidcCache.config, nil
+	}
+	lock.RUnlock()
+
+	var c oidcConfig
+	entry, err := s.Get(ctx, oidcConfigStorageKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry != nil {
+		if err := entry.DecodeJSON(&c); err != nil {
+			return nil, err
+		}
+	}
+
+	c.effectiveIssuer = c.Issuer
+	if c.effectiveIssuer == "" {
+		c.effectiveIssuer = i.core.redirectAddr + issuerPath
```

Comment:
What happens here if i.core.redirectAddr is empty? You just get issuerPath, but...could that cause weird behavior?

Reply:
I think the best we can do is log a warning, here and possibly on startup (e.g. some one-time check in the factory method). I think in most cases (and any case involving discovery), the issuer needs to be a real, accessible address. So if api_addr is "", they should be encouraged to configure Issuer manually.

Reply:
Added warning on config read. If this case is happening, most likely they're not going to get very far with using their tokens.

Reply:
Maybe also only concatenate if the redirectAddr != ""?